python-libraries/nanover-ase/src/nanover/ase/imd_calculator.py

```diff
@@ -10,14 +10,101 @@
 from ase.calculators.calculator import Calculator, all_changes
 from ase.md.md import MolecularDynamics
 from ase.md.velocitydistribution import _maxwellboltzmanndistribution
-from nanover.imd.imd_force import calculate_imd_force
+from nanover.imd.imd_force import calculate_imd_force, get_sparse_forces
 from nanover.imd.imd_state import ImdStateWrapper
 from nanover.imd.particle_interaction import ParticleInteraction
-from nanover.trajectory.frame_data import MissingDataError
+from nanover.trajectory.frame_data import MissingDataError, FrameData
 
 from . import converter
 
 
+class ImdForceManager:
+    """
+    A class that calculates and stores the iMD forces and energies for
+    an ASE simulation.
+    """
+
+    def __init__(self, imd_state: ImdStateWrapper, atoms: Atoms):
+        self.atoms = atoms
+        self.imd_state = imd_state
+
+        self.total_user_energy = 0.0
+        self.user_forces: np.ndarray = np.zeros(self.atoms.positions.shape)
+
+        self._previous_interactions: Dict[str, ParticleInteraction] = {}
+        self.call_reset_velocities: bool = False
+
+    def update_interactions(self):
+        """
+        Update the iMD interaction energies and forces (in ASE units)
+        """
+        self._update_forces(self.atoms)
+
+    def add_to_frame_data(self, frame_data: FrameData):
+        """
+        Add the iMD forces and energy to the frame data
+        """
+        frame_data.user_energy = self.total_user_energy * converter.EV_TO_KJMOL
+        user_sparse_indices, user_sparse_forces = get_sparse_forces(self.user_forces)
+        frame_data.user_forces_sparse = user_sparse_forces * (
+            converter.EV_TO_KJMOL / converter.ANG_TO_NM
+        )
+        frame_data.user_forces_index = user_sparse_indices
+
+    def _update_forces(self, atoms):
+        """
+        Get the forces to apply from the iMD service and communicate them
+        """
+        # Get the current interactions from the iMD service (if any)
+        interactions = self.imd_state.active_interactions
+
+        # Determine whether interactions are the same as previous, and if
+        # not call _reset_velocities in ImdCalculator
+        if interactions != self._previous_interactions:
+            self.call_reset_velocities = True
```

Comment:
I think ditch this check and field and assume it's always true -- reset velocities performs its own checks on if interactions need to be reset or not.

Reply:
On further discussion: set it to True regardless -- it's equivalent to signalling that forces have changed since the last reset.